File: benchmark/elixir/call_bench.exs
Logger.configure(level: :warning)

defmodule CallBench do
  @calls 10_000
  @calls_contention 5_000

  def run_sequential(calls) do
    {:ok, pid} = Stages.Counter.start_link()

    for _ <- 1..calls do
      GenServer.call(pid, :increment)
    end

    GenServer.stop(pid)
  end

  def run_sequential_echo(calls) do
    {:ok, pid} = Stages.Echo.start_link()

    for i <- 1..calls do
      ^i = GenServer.call(pid, i)
    end

    GenServer.stop(pid)
  end

  def run_parallel(calls, num_actors) do
    actors =
      for _ <- 1..num_actors do
        {:ok, pid} = Stages.Echo.start_link()
        pid
      end

    calls_per_actor = div(calls, num_actors)

    for _ <- 1..calls_per_actor do
      actors
      |> Enum.map(fn pid -> Task.async(fn -> GenServer.call(pid, :ping) end) end)
      |> Task.await_many()
    end

    Enum.each(actors, &GenServer.stop/1)
  end

  def run_contention(total_calls, in_flight) do
    {:ok, pid} = Stages.Echo.start_link()

    Stream.chunk_every(1..total_calls, in_flight)
    |> Enum.each(fn batch ->
      batch
      |> Enum.map(fn i -> Task.async(fn -> GenServer.call(pid, i) end) end)
      |> Task.await_many()
    end)

    GenServer.stop(pid)
  end

  def run_chain(calls, depth) do
    {:ok, echo} = Stages.Echo.start_link()

    target =
      Enum.reduce(1..depth, echo, fn _, prev ->
        {:ok, fwd} = Stages.Forwarder.start_link(prev)
        fwd
      end)

    for i <- 1..calls do
      ^i = GenServer.call(target, i)
    end

    GenServer.stop(target)
    if depth > 0, do: GenServer.stop(echo)
  end

  def run_mixed(calls) do
    {:ok, pid} = Stages.Counter.start_link()

    for i <- 1..calls do
      if rem(i, 2) == 0 do
        GenServer.call(pid, :increment)
      else
        GenServer.cast(pid, :increment)
      end
    end

    GenServer.call(pid, :get)
    GenServer.stop(pid)
  end

  def run do
    IO.puts(String.duplicate("=", 60))
    IO.puts("Elixir Call Benchmarks")
    IO.puts("Elixir #{System.version()} on OTP #{System.otp_release()}")
    IO.puts(String.duplicate("=", 60))
    IO.puts("")

    benchee_opts = [time: 3, warmup: 1, print: [configuration: false]]

    Benchee.run(
      %{
        "#{@calls} sequential calls" => fn -> run_sequential(@calls) end,
        "#{@calls} sequential echo calls" => fn -> run_sequential_echo(@calls) end
      },
      [{:title, "Sequential calls: 1 caller → 1 actor"} | benchee_opts]
    )

    Benchee.run(
      for n <- [1, 4, 8, 16], into: %{} do
        {"#{@calls} calls → #{n} actor(s)", fn -> run_parallel(@calls, n) end}
      end,
      [{:title, "Parallel calls: fan-out to N actors"} | benchee_opts]
    )

    Benchee.run(
      for n <- [10, 50, 100], into: %{} do
        {"#{@calls_contention} calls, #{n} in-flight",
         fn -> run_contention(@calls_contention, n) end}
      end,
      [{:title, "Contention: N callers → 1 actor"} | benchee_opts]
    )

    Benchee.run(
      %{
        "#{@calls} calls: A → B (1 hop)" => fn -> run_chain(@calls, 1) end,
        "#{@calls} calls: A → B → C (2 hops)" => fn -> run_chain(@calls, 2) end
      },
      [{:title, "Call chain depth"} | benchee_opts]
    )

    Benchee.run(
      %{
        "#{@calls} ops (50/50 call/cast)" => fn -> run_mixed(@calls) end
      },
      [{:title, "Mixed: calls + casts interleaved"} | benchee_opts]
    )
  end
end

CallBench.run()

File: benchmark/call.bench.ts
import { bench, describe } from "vitest";
import { Actor, ActorRef, ActorSystem } from "../src";
import { createBenchSystem } from "./helpers";

const CALLS = 10_000;
const CALLS_CONTENTION = 5_000;

// Lazy singleton — created once, reused across all benches.
let _system: ActorSystem | null = null;
async function getSystem(): Promise<ActorSystem> {
  if (!_system) _system = await createBenchSystem();
  return _system;
}

const opts = { time: 1000, warmupIterations: 1, warmupTime: 100 };

// ============================================================
//  Actors
// ============================================================

class CounterActor extends Actor<{ type: "increment" }, { type: "get" | "increment" }, number> {
  private count = 0;

  handleCall(message: { type: "get" | "increment" }): number {
    if (message.type === "increment") {
      return ++this.count;
    }
    return this.count;
  }

  handleCast(message: { type: "increment" }): void {
    this.count++;
  }
}

class EchoActor extends Actor<never, any, any> {
  handleCall(message: any): any {
    return message;
  }

  handleCast(): void {}
}

// ============================================================
//  Sequential Calls: One caller → one actor, serial
// ============================================================

describe("Sequential calls: 1 caller → 1 actor", () => {
  bench(
    `${CALLS} sequential calls`,
    async () => {
      const system = await getSystem();
      const ref = system.spawn(CounterActor);
      for (let i = 0; i < CALLS; i++) {
        await ref.call({ type: "increment" });
      }
      await system.stop(ref);
    },
    opts,
  );

  bench(
    `${CALLS} sequential echo calls (pass-through)`,
    async () => {
      const system = await getSystem();
      const ref = system.spawn(EchoActor);
      for (let i = 0; i < CALLS; i++) {
        await ref.call(i);
      }
      await system.stop(ref);
    },
    opts,
  );
});

// ============================================================
//  Parallel Calls: One caller → N actors, parallel batches
// ============================================================

describe("Parallel calls: fan-out to N actors", () => {
  for (const N of [1, 4, 8, 16]) {
    bench(
      `${CALLS} calls → ${N} actor(s), batches of ${N}`,
      async () => {
        const system = await getSystem();
        const actors: ActorRef[] = [];
        for (let i = 0; i < N; i++) {
          actors.push(system.spawn(EchoActor));
        }
        const callsPerActor = Math.floor(CALLS / N);
        for (let i = 0; i < callsPerActor; i++) {
          await Promise.all(actors.map((ref) => ref.call(i)));
        }
        await Promise.all(actors.map((ref) => system.stop(ref)));
      },
      opts,
    );
  }
});

// ============================================================
//  Contention: N callers → 1 actor, all in flight
// ============================================================

describe("Contention: N callers → 1 actor", () => {
  for (const N of [10, 50, 100]) {
    bench(
      `${CALLS_CONTENTION} calls, ${N} in-flight at a time`,
      async () => {
        const system = await getSystem();
        const ref = system.spawn(EchoActor);
        let completed = 0;

        while (completed < CALLS_CONTENTION) {
          const batch = Math.min(N, CALLS_CONTENTION - completed);
          const promises: Promise<any>[] = [];
          for (let i = 0; i < batch; i++) {
            promises.push(ref.call(completed + i));
          }
          await Promise.all(promises);
          completed += batch;
        }

        await system.stop(ref);
      },
      opts,
    );
  }
});

// ============================================================
//  Call chain: A calls B calls C (nested calls)
// ============================================================

class ForwarderActor extends Actor {
  private target!: ActorRef;

  init(target: ActorRef) {
    this.target = target;
  }

  handleCall(message: any): Promise<any> {
    return this.target.call(message);
  }

  handleCast(): void {}
}

describe("Call chain depth", () => {
  bench(
    `${CALLS} calls: A → B (1 hop)`,
    async () => {
      const system = await getSystem();
      const echo = system.spawn(EchoActor);
      const forwarder = system.spawn(ForwarderActor, { args: [echo] });
      for (let i = 0; i < CALLS; i++) {
        await forwarder.call(i);
      }
      await system.stop(forwarder);
      await system.stop(echo);
    },
    opts,
  );

  bench(
    `${CALLS} calls: A → B → C (2 hops)`,
    async () => {
      const system = await getSystem();
      const echo = system.spawn(EchoActor);
      const mid = system.spawn(ForwarderActor, { args: [echo] });
      const front = system.spawn(ForwarderActor, { args: [mid] });
      for (let i = 0; i < CALLS; i++) {
        await front.call(i);
      }
      await system.stop(front);
      await system.stop(mid);
      await system.stop(echo);
    },
    opts,
  );
});

// ============================================================
//  Mixed workload: calls + casts interleaved
// ============================================================

describe("Mixed: calls + casts interleaved", () => {
  bench(
    `${CALLS} ops (50/50 call/cast)`,
    async () => {
      const system = await getSystem();
      const ref = system.spawn(CounterActor);
      for (let i = 0; i < CALLS; i++) {
        if (i % 2 === 0) {
          await ref.call({ type: "increment" });
        } else {
          ref.cast({ type: "increment" });
        }
      }
      // Final call to ensure all casts are flushed
      await ref.call({ type: "get" });
      await system.stop(ref);
    },
    opts,
  );
});
